### PR TITLE
RES_NET_DEV_ALL code - parsing /proc/net/dev

### PR DESCRIPTION
**Pull request by anjalidk**

### CODE REVIEW

File: net_if.c [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+		total_if_size += if_size;
+#ifdef PRINTLOGS
+		printf("max_ifs %d, if_size %d total_if_size %d\n",max_ifs, if_size, total_if_size);
+#endif
+		if (ifaces == NULL) {
+			ifaces = (struct ifstats *) malloc(total_if_size);
+			ifs = ifaces;
+		} else {
+			ifaces = (struct ifstats *)
+				realloc(ifaces, total_if_size);
```
Comment (hallyn) [inline net_if.c:172]: realloc should really be done as
```
void *tmp = realloc(ifaces, total_if_size);
if (!tmp) {
   free(ifaces);
   return -1;
}
ifaces = tmp;

File: net_if.c [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+		iov.iov_base = buf;
+		iov.iov_len = len;
+		max_ifs = get_max_ifaces(len);
+		if_size = max_ifs * sizeof(struct ifstats);
+		total_if_size += if_size;
+#ifdef PRINTLOGS
+		printf("max_ifs %d, if_size %d total_if_size %d\n",max_ifs, if_size, total_if_size);
+#endif
+		if (ifaces == NULL) {
+			ifaces = (struct ifstats *) malloc(total_if_size);
```
Comment (hallyn) [inline net_if.c:168]: Not checking return value

File: net_if.c [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+	msg.msg_iov = &iov;
+	msg.msg_iovlen = 1;
+
+	while(1) {
+		len = get_if_len(net_sock);
+		if (len == -1) {
+			close(net_sock);
+			return -1;
+		}
+		buf = (char *)malloc(len);
```
Comment (hallyn) [inline net_if.c:158]: check for null

File: net_if.c [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+		} else {
+			ifaces = (struct ifstats *)
+				realloc(ifaces, total_if_size);
+			ifs = ifaces + ifind;
+		}
+		recvl = recvmsg(net_sock, &msg, 0);
+		if (recvl < 0) {
+			err = errno;
+			eprintf("Error in response recvmsg(), errno %d\n", err);
+			errno = err;
```
Comment (hallyn) [inline net_if.c:179]: Note you're not freeing buf or ifaces here.

**Comment by anjalidk** [comment]
Done, took care of all review comments and squashed this commit into previous one. Pull request is still open.

Anjali

File: net_if.c [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+	int ifind = 0, total_if_size = 0;
+	struct iovec iov;
+	struct msghdr msg;
+	struct nlmsghdr *r;
+	struct nlmsgerr *nlmsg_err;
+	struct if_stats_msg *ifsm;
+	struct rtattr *at[IFLA_STATS_MAX+1];
+	struct rtattr *rta;
+	struct ifstats *ifaces = NULL, *ifs = NULL;
+	char *buf;
```
Comment (hallyn) [inline net_if.c:144]: It seems better to define char *buf only in the scope of the while(1) loop.

File: net_if.c [hunk 6]
```diff
@@ -0,0 +1,299 @@
...
+			printf("tx_packets %llu\n",ifs->st64.tx_packets);
+			printf("rx_bytes %llu\n",ifs->st64.rx_bytes);
+			printf("tx_bytes %llu\n",ifs->st64.tx_bytes);
+#endif
+			ifs++;
+			ifind++;
+			r = NLMSG_NEXT(r, recvl);
+			count++;
+		}
+		//free(buf);
```
Comment (hallyn) [inline net_if.c:252]: Is this in fact not needed?

**Comment by anjalidk** [comment]
Done, and pushed again.

**Comment by anjalidk** [comment]
> On Jun 5, 2023, at 4:24 PM, Serge Hallyn ***@***.***> wrote:
> 
> 
> @hallyn commented on this pull request.
> 
> In net_if.c:
> 
> > +			if (if_indextoname(ifsm->ifindex, ifs->ifname) == NULL)
> +				eprintf("Error in if_indextoname %d\n",errno);
> +#ifdef PRINTLOGS
> +			else
> +				printf("Interface name %s\n",ifs->ifname);
> +			printf("rx_packets %llu\n",ifs->st64.rx_packets);
> +			printf("tx_packets %llu\n",ifs->st64.tx_packets);
> +			printf("rx_bytes %llu\n",ifs->st64.rx_bytes);
> +			printf("tx_bytes %llu\n",ifs->st64.tx_bytes);
> +#endif
> +			ifs++;
> +			ifind++;
> +			r = NLMSG_NEXT(r, recvl);
> +			count++;
> +		}
> +		//free(buf);
> 
> Is this in fact not needed?
No, not needed, removed it.

**Comment by hallyn** [comment]
So buf gets assigned to the msg.iov.iov_base, and gets freed in most exit paths...  But why does it not need to be freed at the bottom of the while loop at the top of which it gets allocated?

**Comment by anjalidk** [comment]
The correct path exit (when there is no error) is via the condition:
r->nlmsg_type == NLMSG_DONE.
So the last nlmsghdr always has NLMSG_DONE. 
The only reason it may exit from under the while loop is if the nlmsghdr is  malformed or not full present etc.
Anjali

**Comment by hallyn** [comment]
> The correct path exit (when there is no error) is via the condition: r->nlmsg_type == NLMSG_DONE. So the last nlmsghdr always has NLMSG_DONE. The only reason it may exit from under the while loop is if the nlmsghdr is malformed or not full present etc.

Can it go through the outer while(1) loop twice?  If not, then why is it a while(1) loop?  If so, then I believe the first buf allocation will be leaked.

(Ideally this code would use a few well architected helpers so that the main handle_stats_resp() would fit on one page and be easier to track, per Linux CodingStyle).

File: net_if.c [hunk 7]
```diff
@@ -0,0 +1,304 @@
...
+		iov.iov_base = buf;
+		iov.iov_len = len;
+		max_ifs = get_max_ifaces(len);
+		if_size = max_ifs * sizeof(struct ifstats);
+		total_if_size += if_size;
+#ifdef PRINTLOGS
+		printf("max_ifs %d, if_size %d total_if_size %d\n",max_ifs, if_size, total_if_size);
+#endif
+		if (ifaces == NULL) {
+			ifaces = (struct ifstats *) malloc(total_if_size);
```
Comment (hallyn) [inline net_if.c:173]: Thanks, this looks good, one last comment though.  Alex would yell at me if I did not point out that realloc(NULL, size) works as malloc.  So you don't need this whole if (ifaces == NULL) case is not needed, you can just always do tmp = realloc(ifaces, total_if_size); and on the first run, ifaces will be NULL, which is fine.

**Comment by anjalidk** [comment]
> On Jun 19, 2023, at 10:20 AM, Serge Hallyn ***@***.***> wrote:
> 
> 
> @hallyn commented on this pull request.
> 
> In net_if.c:
> 
> > +		char *buf = (char *)malloc(len);
> +		if (!buf) {
> +			if (ifaces)
> +				free(ifaces);
> +			return -1;
> +		}
> +		iov.iov_base = buf;
> +		iov.iov_len = len;
> +		max_ifs = get_max_ifaces(len);
> +		if_size = max_ifs * sizeof(struct ifstats);
> +		total_if_size += if_size;
> +#ifdef PRINTLOGS
> +		printf("max_ifs %d, if_size %d total_if_size %d\n",max_ifs, if_size, total_if_size);
> +#endif
> +		if (ifaces == NULL) {
> +			ifaces = (struct ifstats *) malloc(total_if_size);
> 
> Thanks, this looks good, one last comment though. Alex would yell at me if I did not point out that realloc(NULL, size) works as malloc. So you don't need this whole if (ifaces == NULL) case is not needed, you can just always do tmp = realloc(ifaces, total_if_size); and on the first run, ifaces will be NULL, which is fine.
Ok, done, fixed this and removed malloc.

**Comment by anjalidk** [comment]
Thanks for merging! Submitted for RES_NET_ROUTE_ALL.
Anjali

**Pull request merged**